src/Npgsql/ConnectorPool.cs

```diff
@@ -320,20 +328,22 @@ await Task.WhenAny(tcs.Task, Task.Delay(timeLeft, delayCancellationToken.Token))
                                 }
                                 else
                                 {
-                                    using (cancellationToken.Register(() => tcs.SetCanceled()))
+                                    using (cancellationToken.Register(s => ((TaskCompletionSource<NpgsqlConnector?>)s).SetCanceled(), tcs))
                                         await tcs.Task;
                                 }
                             }
                             else
                             {
                                 if (timeout.IsSet)
                                 {
-                                    var timeLeft = timeout.TimeLeft;
-                                    if (timeLeft <= TimeSpan.Zero || !tcs.Task.Wait(timeLeft))
+                                    var timeLeft = timeout.TimeLeft.TotalMilliseconds;
+                                    if (timeLeft > int.MaxValue)
+                                        throw new ArgumentOutOfRangeException(nameof(timeout), "timeout.TimeLeft.TotalMilliseconds must be smaller than int.MaxValue.");
+                                    if (timeLeft <= 0 || !tcs.Task.Wait((int)timeLeft, cancellationToken))
                                         throw new NpgsqlException($"The connection pool has been exhausted, either raise MaxPoolSize (currently {_max}) or Timeout (currently {Settings.Timeout} seconds)");
                                 }
                                 else
-                                    tcs.Task.Wait();
+                                    tcs.Task.Wait(cancellationToken);
```

Comment:
You wrote:

> The cancellation token commit feels more consistent with the intent of the work being done whether it's being awaited asynchronously or not but I'm entirely ambivalent either way it goes.

Yeah, but unless I'm mistaken there's no way this cancellation token can ever trigger here in the sync path:
* We could not have gotten a ctoken from the user, and
* Timeout isn't implemented via a ctoken as in the async path

If that's the case and I haven't missed anything, then I think it's better not to pass the ctoken in to make the code more clear.

Reply:
I don't get what you mean with runtime check? The assert is on an internal const.

Reply:
Not in your implementation, I usually like to make coherent apis that when I generalize them to be used in multiple projects there is usually only api shape work left. I can revert that commit of course, but it doesn't do any harm right now, it all shortcircuits on `token.CanBeCanceled` if it's always a default token.

Reply:
I'd rather we remove it. When reading the code it gives the mistaken impression that cancellation is possible there (i.e. I'm not worried about perf or anything, just readability in this already pretty complicated component).

Reply:
I'd say that actually improves maintainability "cancellation is always possible" no exceptions based on branching. But I'll revert the commit :)

Reply:
Thanks, I guess it's a matter of taste/opinion - "cancellation is always possible", but not in this actual specific case. When I saw your commit adding the token, I spend a bit of time analyzing the code to try to understand in which case it may trigger - I'd like to save the mental effort from myself or anyone else reading that code.

Anyway thanks for reverting!